### ISSUE TITLE
Test if Hibernate can handle decent amount of pipelines, executions

### ISSUE THREAD

**Issue by tomas-knap** [mff-uk/odcs#209]
1. Test frontend with 200+ pipelines, 1000+ executions
2. Test frontend with 500+ pipelines, 10000+ executions

**Comment by janvojt** [comment]
As discussed at the meeting, this needs to wait until listing tables are paginated in a lazy manner on DB level.

**Comment by tomas-knap** [comment]
Test also with remote database: I suggest you use my remote testing database, generate 5000 entries to the table storing the execution monitor results (just copy one of the records) and then implement it so that it is fast. You don't need to refresh by downloading everything all the time. Just check with a single simple query whether there is a record newer than the last one you are displaying and if yes, download only those.

**Comment by janvojt** [comment]
I created around 600 pipelines and 10600 executions. Web frontend is working smoothly without noticable change in performance as I expected.

Then I tested how the delays will affect performance to see what might happen if the remote host is quite far. I already suspected this will be a problem, because on some screens we use quite many database queries. Each query needs to wait for the previous one, which can produce quite a delay.

I tested using netemul and emulated a delay of 10ms on my loopback interface with the following command:

```
# tc qdisc add dev lo root netem delay 10ms
```

This delay adds up to 20ms roundtrip, and is roughly an equivalent of web client being in Czech republic, Tomcat server being in England, and Virtuoso server being in Czech republic.

As I expected, it affected screens with many queries. PipelineList generates a query for each of the table cells containing info about last pipeline execution. It performed the worst, page load took around 13 seconds. ExecutionMonitor was a little better, loading in 3 seconds.

BTW I made some optimizations in PipelineList 2 days ago. Without these, loads would be much slower (I am guessing 3-5 times slower).

I think we should optimize PipelineList, other pages seem to load in acceptable time. We should consider that the situation I tested is quite extreme, so I believe we can lower the priority to normal?

The optimizations that can be done in PipelineList are quite straightforward. I believe it would be enough to add 3 columns to pipeline table, where we can save the information about last execution. This would need to be updated by backend when pipeline finishes.

**Comment by tomas-knap** [comment]
Thanks. I created issue for pipeline list optimization.

What was the problem in pipeline execution that it took 3seconds to load the page? 

Did you also test executing certain pipeline and observing the logs in the debug view online with all these latencies?

Regarding the problematical parts, even on the localhost, it took quite a while to load exec monitor record detail from exec monitor, 3-7s on average. what was the result in your case?

**Comment by janvojt** [comment]
> What was the problem in pipeline execution that it took 3seconds to load the page?

For each execution there is one query that fetches pipeline name. As I already mentioned before, this could be resolved by specifying association to preload in LazyQueryContainer. However I do not have time to do this this week.

> Did you also test executing certain pipeline and observing the logs in the debug view online with all these latencies?

Yes, everything else was OK. However I did not test debug run, that might slow things down when heavy logging into database. I did not test this, because there is not much we could do with that. Such setup is extreme, and debugging pipeline would be even more extreme.

> Regarding the problematical parts, even on the localhost, it took quite a while to load exec monitor record detail from exec monitor, 3-7s on average. what was the result in your case?

Opening execution detail takes about 13 seconds. I did not notice this before, because I was opening executions with just a few logs (copied ones). Therefore as I found, this is also slow. The problem is, that quite a few things need to be loaded (Pipeline, Graph, Nodes, DPU instances). Also, similar as in PipelineList, for each log, logging property needs to be loaded. Unlike last executions in PipelineList, this could be resolved by specifying association to preload in LazyQueryContainer.
